text/3668-async-closure.md

```diff
@@ -0,0 +1,625 @@
+
+```rust
+/// In return-position impl trait:
+fn closure() -> impl async Fn() { async || {} }
+
+/// In trait bounds:
+trait Foo<F>: Sized
+where
+    F: async Fn()
+{
+    fn new(f: F) -> Self;
+}
+
+/// in GATs:
+trait Gat {
+    type AsyncHasher<T>: async Fn(T) -> i32;
+}
+```
+
+# Detailed Explanation
+
+### `AsyncFn*`
+
+This RFC introduces a family of `AsyncFn` traits. These traits are intended to remain unstable to name or implement, just like the `Fn` traits. Nonetheless, we'll describe the details of these traits so as to explain the user-facing features enabled by them.
+
+The definition of the traits is (modulo `rustc_` attributes, and the `"rust-call"` ABI):
+
+
+> [!NOTE]
+> We omit some details about the `"rust-call"` calling convention and the fact that the `Args` parameter is enforced to be a tuple.
+
+```rust
+/// An async-aware version of the [`FnOnce`](crate::ops::FnOnce) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnOnce::async_call_once`].
+    type CallOnceFuture: Future<Output = Self::Output>;
+
+    /// Output type of the called closure's future.
+    type Output;
+
+    /// Call the [`AsyncFnOnce`], returning a future which may move out of the called closure.
+    fn async_call_once(self, args: Args) -> Self::CallOnceFuture;
+}
+
+/// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnMut<Args>: AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnMut::async_call_mut`] and [`AsyncFn::async_call`].
+    type CallRefFuture<'a>: Future<Output = Self::Output>
+    where
+        Self: 'a;
+
+    /// Call the [`AsyncFnMut`], returning a future which may borrow from the called closure.
+    fn async_call_mut(&mut self, args: Args) -> Self::CallRefFuture<'_>;
+}
+
+/// An async-aware version of the [`Fn`](crate::ops::Fn) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFn<Args>: AsyncFnMut<Args> {
+    /// Call the [`AsyncFn`], returning a future which may borrow from the called closure.
+    fn async_call(&self, args: Args) -> Self::CallRefFuture<'_>;
+}
+```
+
+### Associated types of `AsyncFn*` traits are not nameable
+
+Unlike what is true today with the current `Fn*` traits, this RFC reserves as an implementation detail the associated types of the `AsyncFn*` traits, and these will not be nameable as part of the stable interface specified by this RFC.
+
+That is, using the existing `FnOnce` trait, we can write this today on stable Rust:
+
+```rust
+fn foo<F, T>()
+where
+    F: FnOnce() -> T,
+    F::Output: Send, //~ OK
+{
+}
+```
+
+(We decided to allow this in [#34365](https://github.com/rust-lang/rust/pull/34365).)
+
+However, this RFC reserves as an implementation detail the associated types of the traits specified above, so this does not work:
```

Comment:
The problem here is if we change `AsyncFnOnce` to desugar to an `FnOnce` trait goal with an associated `Future` type, i.e. change `where F: async FnOnce() -> T` to desugar to `where F: FnOnce<Output: Future<Output = T>>`, then the meaning of `T::Output` changes. I don't want to have to add another hack into the language and compiler to preserve the meaning of `::Output` here, especially since the only real argument for it is about consistency.

 It can be stabilized separately if there's a good reason to, and that stabilization will have the burden of proving why `Output` is useful.

Reply:
> [...] especially since the only real argument for it is about consistency.

I certainly believe that "consistency" is an acceptable, self-contained argument. I would like for Rust to have a consistent user experience, and for that I believe we need to be able to argue for consistency to achieve that. Otherwise we're in a position where each similar API may find itself re-litigating API decisions on a per-case basis, and I don't think that would be good.

As you're stating here, we're now leaving a gap open in the design about what the value of `Output` should be. When we have a `F: async FnOnce() -> T`, what should `F::Output` evaluate to?:

- `T`
- `impl Future<Output = T>`.

I believe the latter would be incredibly confusing for people. I believe this RFC should make a decision on what the value of `F::Output` should be. And if we're deciding anyway, we should probably also implement that. This is definitely not something we should punt.

Reply:
I don't think you've demonstrated why this decision needs to be made now, and I disagree that consistency is compelling enough to possibly lock us into dealing with migrating the meaning of `::Output` later, so I'm probably just going to summarize this discussion and leave it in the "alternatives" or "future" section.

Reply:
> As you're stating here, we're now leaving a gap open in the design about what the value of Output should be.

I just wanted to repeat that, yes, we are choosing to leave this question open because I don't see any reason for users to need to name `F::Output` in their APIs when their trait bounds already constrain that output to another type is necessarily already nameable. It's the same reason that we're reserving the *whole* definition of the `AsyncFn*` traits open -- because users only need to interact with them via `async Fn*()` trait bounds to realize their power.

There's not really much of a consistency argument to be made here, though. The reason that `Future::Output` is stable is because users need to be able to specify the associated type, and `Future` trait bounds are not like `Fn` trait bounds in the way that they force you to constrain their output type with the `-> U`.

This wasn't done to be consistent with the other traits in `std::ops`, as far as I can tell. I'm gonna repeat myself and say that the only reason that `FnOnce::Output` was stabilized was because of a compiler bug and a "I don't see why not", not because it gives users any more power over their APIs.

 > what should F::Output evaluate to?: * impl Future<Output = T>.

This is a really hard question to answer right now. If we make it resolve to `T`, then any future definition of `async Fn() -> T` will need to preserve that behavior, probably at the price of by adding yet another resolver or lowering wart to the compiler to support code that couldn't just be codemod'ed away to not use that associated type and instead just name `T` directly. And it can't resolve to the `CallOnceFuture`, since bounds on that type won't apply to the `CallRefFuture` associated types that would be present on the `async Fn`/`async FnMut` bounds -- and those types are already boundable via RTN in a way that is less dependent on the specifics of the name resolution of associated types.

Reply:
Specifically, the "I don't see why not to stabilize FnOnce::Output" argument can't really be applied by analogy here , and that's  important -- because unlike `FnOnce` which really doesn't have more than one answer to this question, `AsyncFnOnce` *does* have a choice, and this make that question interesting enough to let it sit out of initially RFCing and stabilizing async fn bounds and async closures.

It seems very premature to have to answer this question right now, and I don't really see what it buys us.

Reply:
> I'm gonna repeat myself and say that the only reason that FnOnce::Output was stabilized was because of a https://github.com/rust-lang/rust/pull/34365 and a https://github.com/rust-lang/rust/pull/34365#issuecomment-229794689, not because it gives users any more power over their APIs.

Yes, I understand that. However, it's now almost a decade later and `::Output` has been part of the public API for the last 60 odd Rust releases. There is now a clear precedent for providing an associated `::Output` type on functions, which necessarily couldn't have been part of the argument then.

Think of this less as consistency for consistency's sake - but instead as an argument for portability. Just like we want to share APIs between, say, Windows and Linux - we also want APIs to be consistent between const and non-const. And between async and non-async. Especially if we're interested in eventually enabling the compiler to automatically abstract between API variants.

It's little inconsistencies like these that will end up making up the long tail of fixes and if enough inconsistencies creep in, we may not even be able to address them. That's why I'm advocating for addressing those as we uncover them, and not punt decisions until some unspecified later date. Uncovering and working through unexpected implications like these is also why we have an RFC process.

> [...] unlike FnOnce which really doesn't have more than one answer to this question, `AsyncFnOnce` does have a choice, and this make that question interesting enough to let it sit out of initially RFCing and stabilizing async fn bounds and async closures.

What worries me about the phrasing here is that this leaves open for interpretation what this feature actually is. This RFC should be describing async closures as: *"The `async/.await` version of Rust's closures"*. When users hear the phrase "async closures" that would be a reasonable expectation based on their experience with Rust's other async features. Especially since this the way this feature works will end up reinforcing that belief.

Because async closures behave like the `async/.await` version of non-async closures, users will reasonably intuit that to be backed by the `async/.await` version of the synchronous closure traits. Now, both traits are unstable so this actually isn't true today. But the expectation will be that if a synchronous closure trait looks like this:

```rust
pub trait FnOnce<Args: Tuple> {
    type Output;
    fn call_once(self, args: Args) -> Self::Output;
}
```

Then the asynchronous counterpart to that trait should look like this:

```rust
pub trait AsyncFnOnce<Args: Tuple> {
    type Output;
    async fn call_once(self, args: Args) -> Self::Output;
}
```

Now, that isn't true today - and not even the synchronous traits are stable yet. But I can say I'm very uncomfortable with the idea that somehow we might later decide, intentionally or not, that the associated type `Output` here may actually begin returning `-> impl Future<Output = T>` rather than a type `T`. That would break the mental model people have of the way this feature works.

**Irrespective of whether we choose to expose the associated `Output` type today; I believe that at a minimum this RFC should codify that the associated type `Output` here should always resolve to type `T` and never to an `impl Future<Output = T>`.** Can we please raise this specific point as a concern on this RFC?

Reply:
I'm realizing I basically just posted an essay-length comment. To maybe more briefly relay my sentiment here: I actually care less about exposing `Output` straight away than I care about codifying that, if exposed, `::Output` should always resolve to `T` and never to `impl Future<Output = T>`.

I don't think that should warrant a particularly big discussion. But it is a decision about the language, so I think we should just decide on it as part of this RFC.

Reply:
I don't see why this RFC has to decide that question, unless there's a compelling case where `F::Output` is actually needed. A commitment that doesn't even have any consequences observable on stable today seems kind of pointless. In fact I see it as a strength of this RFC that it does *not* have to decide on whether `AsyncFnOnce` is just `FnOnce` plus the async effect -- as you know there's no consensus on how the effects story should work around `async`, so it seems premature to commit to anything in this space.

You're saying we should permanently exclude the desugaring [described here](https://github.com/compiler-errors/rust-rfcs/blob/async-closure-redux/text/3668-async-closure.md#why-do-we-need-a-new-set-of-asyncfn-traits). This does seem like a non-trivial decision to me, and there's no strong reason to make that decision right now.

There are also ways in which having `Output` be `T` and not the future is strangely inconsistent. For instance, with RTN, the "return type" of an `async fn` in a trait is considered to be the future -- that's where the `Send` bound applies to. But then with async closures, where `Output` is the name for the return type, wouldn't it be strange for the return type here to not be the closure? I am sure your answer is "no that's not strange at all", and that's fine, all I am pointing out is that this is a commitment that can easily lead to big discussions, and it's entirely unnecessary to have these discussions right now.

Reply:
> To maybe more briefly relay my sentiment here: I actually care less about exposing Output straight away than I care about codifying that, if exposed, `::Output` should always resolve to T and never to `impl Future<Output = T>`.

The RFC **does** takes a stance on this question, specifically stating that that definition of `Output` is **not stable**, and this is the only stance that's practical to take at this point.

There is no benefit to decide on the meaning out `Output` unless we plan to stabilize it, which we **do not** plan on doing. RFCs are in practice not prescriptive in the way that you are implying that they are. As the design of a feature evolves, a future stabilization report is free to change the decisions of an RFC (rust-lang/rust#122055 literally has a section named **"how does this differ from the RFC?"** because the RFC was so wrong in specifics it prescribed on the desugarings of associated type bounds, e.g. in associated type item bounds), so this does not buy us a one-way ticket to exclude the definition of `Output` that you see as a hazard in the future.

Even if it did, and we found some way to legally bind the definition of `Output` to `T` in a way that holds *even though* we are not stabilizing the definition of any of these traits, I agree with @RalfJung above that it is a strength of this RFC that it does not decide these details right now. Ralf is correct in noting that there's no consensus on whether `async FnOnce` will literally become an `async`-keyword-modified `FnOnce` trait or *what* will happen to the trait in the future -- none of the stable surface of this RFC precludes or requires that. However, I'm not convinced that we'd be doing anything other than being incredibly pedantic or unnecessarily locking ourselves into a future `async`-trait-bound-modifier modality by saying that `Output` should forevermore be prescribed to evaluate to the type returned by awaiting the future of the async closure (or at least forcing ourselves to dig ourself out of that hole by having to add hacks to the resolver or language, if the trait ends up diverging in the other direction).

I feel like a broken record here repeating myself, but this should not be done. As I said before, this is a premature decision that I **chose not** to do in this RFC. I think at this point we've gone back and forth enough to make it clear that we just have to agree to disagree.

> I don't think that should warrant a particularly big discussion.

I don't particularly care about the size of the decision. I think it's clear that you think that this is a simple and obvious detail, but I don't. I think it is a detail that does not improve the RFC in a material way given that we have already carved the decision out as irrelevant by reserving the *whole* definition as unstable.

I hope it's clear that the only reason I elaborated anything about these `AsyncFn*` traits in this RFC (which remain unstable!!) is because I needed to point at specifics like GATs to explain things using concrete definitions. I know that's weird, process-wise, to even discuss the definition of things that are kind of "fake" to the user-facing surface of the language, but I literally needed these definitions to answer the even more trivial question of "why do we even need any of this in the first place". However, I honestly kind of regret the fact that I had to provide these definitions at all, since I think that this conversation is not doing anything to improve the RFC's *actual* meaningful, user-facing parts! If there were some other way to have described the way that `async Fn*` bounds and async closures worked uniquely without giving specific trait definitions, I would've definitely done that instead to avoid this whole sidebar. I can't change that now.

> But it is a decision about the language, so I think we should just decide on it as part of this RFC.

This is not a decision about the language. The definition and naming of associated types are a library detail. This RFC is not a T-libs-api RFC because we do not stabilize any details of the trait definitions.

Reply:
> You're saying we should permanently exclude the desugaring [described here](https://github.com/compiler-errors/rust-rfcs/blob/async-closure-redux/text/3668-async-closure.md#why-do-we-need-a-new-set-of-asyncfn-traits).

No, that's not quite right. Note that this only becomes relevant if we subsequently also follow through on a decision to then actually *expose* `Output`. There are two separate questions at play here: should `Output` be exposed at all? And if it is, what type `Output` should resolve into. I believe it's fine to defer the former as that is not a lang decision. But the latter is a lang decision, and I believe should be decided by this RFc.

Also, it's worth adding: even if we decided in this RFC to guarantee `Output = T`, then that still does not preclude changing the underlying desugaring to be based on some form of `LendingFn` trait. As mentioned in this thread: it is possible to special-case the resolution. That might not be the ideal, but that is not the same as being "permanently excluded".

---

> There are also ways in which having `Output` be `T` and not the future is strangely inconsistent. For instance, with RTN, the "return type" of an `async fn` in a trait is considered to be the future -- that's where the `Send` bound applies to. But then with async closures, where Output is the name for the return type, wouldn't it be strange for the return type here to not be the future?

It's worth differentiating here between _concrete return types_ and _logical return types_. Take for example the `IntoFuture` trait, which itself is analogous to `impl AsyncFnOnce() -> T` (async closure with no arguments and a logical return type `T`). It is defined as follows:

```rust
pub trait IntoFuture {
    type Output;
    type IntoFuture: Future<Output = Self::Output>;
    fn into_future(self) -> Self::IntoFuture;
}
```

The _concrete_ output type here is the associated type `IntoFuture`. That's the actual future that we return when we synchronously invoke the function. However, once we evaluate it by invoking `.await`, we're left with the _logical_ return type, which is what the associated `Output` type refers to. The reason why we've added RTN is because the `async fn` notation only describes the logical return type in the signature (`::Output`), and we need RTN to be able to name the concrete return type too (`impl Future<Output = T>`).

| | `async fn() -> T` | `impl async Fn() -> T`| `IntoFuture<Output = T>` |
|---|--- | --- | --- |
| `::Output` | `T` | `T` | `T` |
| `::(..)` | `impl Future<Output = T>` | `impl Future<Output = T>` | `::IntoFuture: Future<Output = T>` |

The same rationale should be applied to async closures too. `::Output` corresponds to the _logical_ return type of the closure (`T`), and RTN should be used to be able to name the _concrete_ return type (`impl Future<Output = T>`). Of course I might be missing something, but this currently does not seem ambiguous to me?

Reply:
It's forward compatible to add `::TheFuture` support to `AsyncFn*` later. It is unclear what the right design here is, and `Fn*::Output` being stable was an accident that was then explicitly stabilized to avoid breaking changes. We can either postpone this question, or we can block the RFC stabilization on it indefinitely.

Let's postpone and discuss this separately as it's just not necessary for an MVP. Doing incremental steps is better than doing no steps.

Reply:
> When we have an `F: async FnOnce() -> T`, what should `F::Output` evaluate to?:
>
> - T
> - impl Future<Output = T>.
>
> I believe the latter would be incredibly confusing for people.

IMO, *both* options would be quite confusing and inconsistent. If we choose the first option, `AsyncFnOnce::Output` differs from `AsyncFnOnce::Output::Output`; if we choose the second, the `Output` of an `AsyncFnOnce` differs from that of a semantically identical `FnOnce`. I think the core issue is that `FnOnce::Output` is badly named, should have been named `FnOnce::Return`; and we should consider renaming it (keeping `Output` as a deprecated alias). That way, `AsyncFnOnce::Return` can be the returned future, and `AsyncFnOnce::Output` (not deprecated) can be the future output type. However, if a rename is off the table, I actually think the second option is least bad

Reply:
Clearly emotions are running a little high in this thread - and I'm sorry if my comments here have caused any stress. It is obviously not my intention to be perceived as engaging in pedantry. But as I've mentioned before: I do feel a responsibility towards the Rust project to, in my capacity as a WG Async member, do my best to be thorough and attentive when it comes to anything async-related. As always though: fine lines to walk, etc. And I might have gotten the balance wrong here. Sorry about that.

Anyway, I've just come back from a walk; had a coffee, touched grass. I just now finished re-reading the thread. And I think this earlier section by @compiler-errors errors is actually rather illuminating. I don't think I fully internalized this on my first read-through:

> I just wanted to repeat that, yes, we are choosing to leave this question open because I don't see any reason for users to need to name `F::Output` in their APIs when their trait bounds already constrain that output to another type is necessarily already nameable. It's the same reason that we're reserving the _whole_ definition of the `AsyncFn*` traits -- because users only need to interact with them via `async Fn*()` trait bounds to realize their power.

By implication, this rationale would then also be extended to any future `LendingFn*` traits (!). Meaning that if anyone wanted to expose `LendingFn*::Output`, they would need to provide a similar motivating use case. At which point T-lang should be involved because it modifies a `#[lang]` item. And possibly WG Async too, because that would have implications for async features. But because it's believed that that seems unlikely to be a thing we would want for any formulation of the backing traits, there's not really any point in discussing that now. Did I get that right?

To be clear: while it might seem like I'm asking about something obvious here, this wasn't obvious to me. I'm legitimately asking whether I've understood this correctly so we can mutually consider this topic resolved.

Reply:
> I don't see any reason for users to need to name `F::Output` in their APIs when their trait bounds already constrain that output to another type is necessarily already nameable.

I totally agree with this, and the decision the RFC makes to keep `Output` unstable. *That being said*, I think the “their trait bounds already constrain that output to another type is necessarily already nameable” is something we will want to change in the future, in order to fix https://github.com/rust-lang/rust/issues/34162#issuecomment-1500599928, at which point naming `F::Output` *will* become useful.

Reply:
I would definitely expect that t-lang and wg-async will be involved before `AsyncFn*::Output` (or any of the other associated types here) becomes accessible on stable.

Reply:
> [...] That being said, I think the “their trait bounds already constrain that output to another type is necessarily already nameable” is something we will want to change in the future [...]

A solution to rust-lang/rust#34162 that relies on bounding an unconstrained `FnOnce::Output` to avoid the higher-ranked-type problem has many other issues to think about, such as closure signature inference, the fact that you cannot name a higher-ranked associated type using a ["type-dependent path"](https://godbolt.org/z/1rTGhfr9x) (which means we need to qualify the type with a trait, which implies we need to decide how to name an `Fn` trait's generics, which is its own beast of a decision), and the fact that changing the signature of an libstd function will need to remove one of its generics (the `B` param) in a backwards-compatible way.

I'm sure that any solution along those lines will require intensive design work and an eventual FCP and is a more appropriate place to consider this question, and the relevant teams (libs/lang/async/types/whatever) can weigh in about the meaning of `AsyncFnOnce::Output` there 😊

Reply:
Thank you all the clarifications - please consider all concerns I've raised in this thread to be resolved ^^